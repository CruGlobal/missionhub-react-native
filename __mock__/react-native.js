import { NativeModules } from 'react-native';

// From https://github.com/facebook/react-native/issues/18279#issuecomment-374177940
// Will be made unnecessary once https://github.com/facebook/react-native/pull/18718 is released. It doesn't seem to be in 0.55.3
NativeModules.BlobModule = {
  ...NativeModules.BlobModule,
  addNetworkingHandler: jest.fn(),
};
