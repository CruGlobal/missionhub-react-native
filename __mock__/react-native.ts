jest.mock(
  'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
  () => ({
    // @ts-ignore
    ...jest.requireActual(
      'react-native/Libraries/Utilities/NativePlatformConstantsIOS',
    ),
    getConstants: () => ({
      forceTouchAvailable: false,
      interfaceIdiom: 'en',
      isTesting: false,
      osVersion: 'ios',
      reactNativeVersion: { major: 60, minor: 1, patch: 0 },
      systemName: 'ios',
    }),
  }),
);
