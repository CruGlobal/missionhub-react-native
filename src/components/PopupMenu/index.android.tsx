import React, { Component } from 'react';
import PropTypes from 'prop-types';
// @ts-ignore
import Menu, { MenuItem } from 'react-native-material-menu';

import KebabIcon from '../../../assets/images/kebabIcon.svg';
import { Touchable } from '../common';
import { TouchableIOSProps } from '../Touchable/index.ios';
import { TouchableAndroidProps } from '../Touchable/index.android';

import styles from './styles';

interface PopupMenuProps {
  title?: string;
  actions: {
    text: string;
    onPress: () => void;
    destructive?: boolean;
  }[];
  buttonProps?: Partial<TouchableIOSProps> & Partial<TouchableAndroidProps>;
  disabled?: boolean;
  triggerOnLongPress?: boolean;
  iconColor?: string;
  testID?: string;
}

// Android only component
class PopupMenu extends Component<PopupMenuProps> {
  // @ts-ignore
  ref = c => (this.menu = c);

  // @ts-ignore
  showMenu = () => this.menu.show();

  // @ts-ignore
  hideMenu = () => this.menu.hide();

  // @ts-ignore
  renderItem = ({ text, onPress }) => {
    const handlePress = () => {
      this.hideMenu();
      onPress();
    };

    return <MenuItem onPress={handlePress}>{text}</MenuItem>;
  };

  renderMenuTrigger = () => {
    const {
      children,
      disabled,
      triggerOnLongPress,
      buttonProps = { style: undefined },
    } = this.props;

    return children ? (
      <Touchable
        disabled={disabled}
        {...buttonProps}
        {...(triggerOnLongPress
          ? { onLongPress: this.showMenu }
          : { onPress: this.showMenu })}
        testID="popupMenuButton"
      >
        {children}
      </Touchable>
    ) : (
      <Touchable
        testID="popupMenuButton"
        disabled={disabled}
        {...buttonProps}
        {...(triggerOnLongPress
          ? { onLongPress: this.showMenu }
          : { onPress: this.showMenu })}
        style={[styles.container, buttonProps.style]}
      >
        <KebabIcon />
      </Touchable>
    );
  };

  render() {
    // @ts-ignore
    const { actions } = this.props;

    return (
      <Menu ref={this.ref} button={this.renderMenuTrigger()}>
        {actions.map(this.renderItem)}
      </Menu>
    );
  }
}

// @ts-ignore
PopupMenu.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      onPress: PropTypes.func.isRequired,
    }),
  ).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element),
    PropTypes.element,
  ]),
  disabled: PropTypes.bool,
  triggerOnLongPress: PropTypes.bool,
  buttonProps: PropTypes.object,
  iconProps: PropTypes.object,
};

export default PopupMenu;
