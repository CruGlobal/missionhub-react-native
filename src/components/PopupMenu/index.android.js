import React, { Component } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

import { showMenu } from '../../utils/common';
import { IconButton } from '../common';

import styles from './styles';

// Android only component
class PopupMenu extends Component {
  handlePress = () => {
    showMenu(this.props.actions, this.menu);
  };

  ref = c => (this.menu = c);

  render() {
    const { containerStyle, iconStyle } = this.props;

    return (
      <View style={[styles.container, containerStyle]}>
        <IconButton
          ref={this.ref}
          name="moreIcon"
          type="MissionHub"
          style={[styles.icon, iconStyle]}
          onPress={this.handlePress}
        />
      </View>
    );
  }
}

PopupMenu.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      onPress: PropTypes.func.isRequired,
    }),
  ).isRequired,
  style: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.number,
    PropTypes.array,
  ]),
};

export default PopupMenu;
