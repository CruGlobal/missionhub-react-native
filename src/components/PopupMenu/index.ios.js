import React, { Component } from 'react';
import { ActionSheetIOS } from 'react-native';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';

import { Touchable, Icon } from '../common';
import { isFunction } from '../../utils/common';

import styles from './styles';

// iOS only component
@translate()
class PopupMenu extends Component {
  open = () => {
    const { t, actions } = this.props;
    const options = actions.map(o => o.text).concat(t('cancel'));

    ActionSheetIOS.showActionSheetWithOptions(
      { options, cancelButtonIndex: options.length - 1 },
      buttonIndex => {
        if (actions[buttonIndex] && isFunction(actions[buttonIndex].onPress)) {
          actions[buttonIndex].onPress();
        }
      },
    );
  };

  render() {
    const { iconProps = {}, style } = this.props;
    return (
      <Touchable onPress={this.open} style={[styles.container, style]}>
        <Icon
          name="moreIcon"
          type="MissionHub"
          {...iconProps}
          style={[styles.icon, iconProps.style]}
        />
      </Touchable>
    );
  }
}

PopupMenu.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      onPress: PropTypes.func.isRequired,
    }),
  ).isRequired,
  iconProps: PropTypes.object,
  style: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.number,
    PropTypes.array,
  ]),
};

export default PopupMenu;
