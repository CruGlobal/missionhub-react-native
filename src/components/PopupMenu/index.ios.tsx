import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';
import { ActionSheetIOS } from 'react-native';

import { IconButton, Touchable } from '../common';
import { isFunction } from '../../utils/common';

import styles from './styles';

// iOS only component
// @ts-ignore
@withTranslation()
class PopupMenu extends Component {
  showMenu = () => {
    // @ts-ignore
    const { actions, t, title } = this.props;

    // @ts-ignore
    const options = actions.map(a => a.text).concat(t('cancel'));
    // @ts-ignore
    const select = i =>
      actions[i] && isFunction(actions[i].onPress) && actions[i].onPress();

    // @ts-ignore
    let destructiveButtonIndex = actions.findIndex(o => o.destructive);
    if (destructiveButtonIndex < 0) {
      destructiveButtonIndex = undefined;
    }

    const params = {
      options,
      cancelButtonIndex: options.length - 1,
      destructiveButtonIndex,
      ...(title ? { title } : {}),
    };

    ActionSheetIOS.showActionSheetWithOptions(params, btnIndex =>
      select(btnIndex),
    );
  };

  render() {
    const {
      children,
      // @ts-ignore
      disabled,
      // @ts-ignore
      triggerOnLongPress,
      // @ts-ignore
      buttonProps = {},
      // @ts-ignore
      iconProps = {},
    } = this.props;

    return children ? (
      <Touchable
        disabled={disabled}
        {...buttonProps}
        {...(triggerOnLongPress
          ? { onLongPress: this.showMenu }
          : { onPress: this.showMenu })}
        testID="popupMenuButton"
      >
        {children}
      </Touchable>
    ) : (
      <IconButton
        type="MissionHub"
        disabled={disabled}
        {...buttonProps}
        {...iconProps}
        {...(triggerOnLongPress
          ? { onLongPress: this.showMenu }
          : { onPress: this.showMenu })}
        name="moreIcon"
        testID="popupMenuButton"
        buttonStyle={[styles.container, buttonProps.style]}
        style={[styles.icon, iconProps.style]}
        onPress={this.showMenu}
      />
    );
  }
}

// @ts-ignore
PopupMenu.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      onPress: PropTypes.func.isRequired,
      destructive: PropTypes.bool,
    }),
  ).isRequired,
  children: PropTypes.element,
  disabled: PropTypes.bool,
  triggerOnLongPress: PropTypes.bool,
  title: PropTypes.string,
  buttonProps: PropTypes.object,
  iconProps: PropTypes.object,
};

export default PopupMenu;
