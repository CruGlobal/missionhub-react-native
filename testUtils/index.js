import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
