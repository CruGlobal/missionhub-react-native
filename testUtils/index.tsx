import React, { ReactElement, ReactNode } from 'react';
import 'react-native';
import { Provider } from 'react-redux';
import { Provider as ProviderLegacy } from 'react-redux-legacy';
import thunk from 'redux-thunk';
import configureStore, { MockStore } from 'redux-mock-store';
// eslint-disable-next-line import/named
import { NavigationParams, NavigationProvider } from 'react-navigation';
import { ApolloProvider } from '@apollo/react-hooks';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { render, RenderAPI } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';
import Enzyme, { shallow as enzymeShallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IMocks } from 'graphql-tools';

import { createApolloMockClient } from './apolloMockClient';
import { createNavigationProp } from './navigationHelpers';

Enzyme.configure({ adapter: new Adapter() });

export const createThunkStore = configureStore([thunk]);

interface RenderWithContextParams {
  initialState?: {};
  store?: MockStore;
  navParams?: NavigationParams;
  mocks?: IMocks;
  noWrappers?: boolean;
}

// Inspiration from https://github.com/kentcdodds/react-testing-library/blob/52575005579307bcfbe7fbe4ef4636147c03c6fb/examples/__tests__/react-redux.js#L69-L80
export function renderWithContext(
  component: ReactElement,
  {
    initialState,
    store = createThunkStore(initialState),
    navParams,
    mocks: mocks = {},
    noWrappers = false,
  }: RenderWithContextParams = {},
) {
  const mockApolloClient = createApolloMockClient(mocks);

  const navigation = createNavigationProp(navParams);

  let renderResult: RenderAPI;
  if (noWrappers) {
    renderResult = render(React.cloneElement(component, { navigation }));
  } else {
    // Warning: don't call any functions in here that return new instances on every call. All the props need to stay the same otherwise rerender won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
      <NavigationProvider value={navigation}>
        <ProviderLegacy store={store}>
          <Provider store={store}>
            <ApolloProvider client={mockApolloClient}>
              {children}
            </ApolloProvider>
          </Provider>
        </ProviderLegacy>
      </NavigationProvider>
    );

    renderResult = render(React.cloneElement(component, { navigation }), {
      wrapper,
    });
  }

  let storedSnapshot: ReactTestRendererJSON | null;
  return {
    ...renderResult,
    store,
    snapshot: () => expect(renderResult.toJSON()).toMatchSnapshot(),
    recordSnapshot: () => (storedSnapshot = renderResult.toJSON()),
    diffSnapshot: () => {
      if (!storedSnapshot) {
        throw new Error(
          'You must call recordSnapshot to store an initial snapshot before calling diffSnapshot',
        );
      }
      expect(
        snapshotDiff(storedSnapshot, renderResult.toJSON()),
      ).toMatchSnapshot();
    },
  };
}

// TODO: Remove all legacy rendering functions below

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (component: ReactElement) => {
  const { toJSON } = render(component);
  expect(toJSON()).toMatchSnapshot();
};

export const renderShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  let renderedComponent = enzymeShallow(
    <ProviderLegacy store={store}>{component}</ProviderLegacy>,
  ).dive();

  // If component has translation wrappers, dive deeper
  while (
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    ((renderedComponent.type() as any).displayName || '').startsWith(
      'withI18nextTranslation(',
    )
  ) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (
  component: ReactElement,
  store = createThunkStore(),
) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};
