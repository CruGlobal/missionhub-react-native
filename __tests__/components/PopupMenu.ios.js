import React from 'react';
import { ActionSheetIOS } from 'react-native';

import PopupMenu from '../../src/components/PopupMenu/index.ios.js';
import { testSnapshotShallow, renderShallow } from '../../testUtils';

const onPress = jest.fn();
let props = {
  actions: [{ text: 'test', onPress }],
};

describe('PopupMenu iOS', () => {
  it('renders with 1 button', () => {
    testSnapshotShallow(<PopupMenu {...props} />);
  });
  it('renders with icon props', () => {
    testSnapshotShallow(<PopupMenu {...props} iconProps={{ size: 24 }} />);
  });

  it('calls the action sheet for ios', () => {
    ActionSheetIOS.showActionSheetWithOptions = jest.fn();
    const instance = renderShallow(<PopupMenu {...props} />).instance();
    instance.open();
    expect(ActionSheetIOS.showActionSheetWithOptions).toHaveBeenCalled();
  });
});
