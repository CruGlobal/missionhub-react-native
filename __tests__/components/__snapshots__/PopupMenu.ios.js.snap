// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupMenu iOS renders with 1 button 1`] = `
<TouchableOpacity
  accessibilityTraits="button"
  activeOpacity={0.6}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Icon
    name="moreIcon"
    style={
      Array [
        Object {
          "color": "#B2B0B2",
          "paddingHorizontal": 10,
        },
        undefined,
      ]
    }
    type="MissionHub"
  />
</TouchableOpacity>
`;

exports[`PopupMenu iOS renders with icon props 1`] = `
<TouchableOpacity
  accessibilityTraits="button"
  activeOpacity={0.6}
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
      undefined,
    ]
  }
>
  <Icon
    name="moreIcon"
    size={24}
    style={
      Array [
        Object {
          "color": "#B2B0B2",
          "paddingHorizontal": 10,
        },
        undefined,
      ]
    }
    type="MissionHub"
  />
</TouchableOpacity>
`;
